hw/ip/otbn/util/shared/control_flow.py

```diff
@@ -386,6 +395,47 @@ def _populate_control_graph(graph: ControlGraph, program: OTBNProgram,
                                          program.max_pc()), [ImemEnd()])
     return
 
+def _label_cycles(program: OTBNProgram, graph: ControlGraph, start_pc: int, visited_pcs: Set[int]) -> None:
+    '''Creates Cycle edges to remove cyclic control flow from the graph.
+
+    Modifies graph in place. Works by replacing edges that loop back to an
+    already-traversed part of the control flow with a Cycle instance.
+    '''
+    sec, edges = graph.get_entry(start_pc)
+    for pc in sec:
+        visited_pcs.add(pc)
+    for i in range(len(edges)):
+        edge = edges[i]
+        if isinstance(edge, Ret) or isinstance(edge, ImemEnd) or isinstance(edge, Ecall):
+            # Cannot possibly loop back 
+            continue
+        elif isinstance(edge, Cycle):
+            # Done; no need to traverse or replace
+            continue
+        if edge.pc in visited_pcs:
+            # Replace with Cycle instance and do not traverse
+            edges[i] = Cycle(edge.pc)
+        else:
+            _label_cycles(program, graph, edge.pc, visited_pcs.copy())
+
+
+def _fix_cycles(program: OTBNProgram, graph: ControlGraph) -> None:
+    '''Labels cycles and splits them from other CodeSections in the graph.
+
+    Modifies graph in place.
+    '''
+    _label_cycles(program, graph, graph.start, set())
+    cycle_start_pcs = graph.get_cycle_starts()
+    new_entries: Dict[int,Tuple[CodeSection,List[ControlLoc]]] = {}
```

Comment:
Does this actually need the type annotation? I would have expected mypy to get it right by itself by looking at line 435?

Reply:
But this is a typing annotation, which is checked by a static type checker... So it's either a no-op (at runtime) or it's not a type unsafe language (when running mypy).

Reply:
Ah, cool. Either way is fine: I only asked because it looked like it had been awkward to write! About the inference: maybe a previous version of the code didn't actually populate the dict. That's the way I generally get tripped up.